// scryfall-client/dist/lib/api-request/send-request-to-api.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var wrap_scryfall_response_1 = __importDefault(require("../../lib/wrap-scryfall-response"));
var get_url_1 = __importDefault(require("./get-url"));
var enque_task_1 = __importDefault(require("./enque-task"));
var send_request_1 = __importDefault(require("./send-request"));
var scryfall_error_1 = __importDefault(require("../../models/scryfall-error"));
function sendRequestToApi(options
// eslint-disable-next-line @typescript-eslint/no-explicit-any
) {
    var url = get_url_1.default(options.endpoint, options.query);
    return enque_task_1.default(function () {
        return send_request_1.default({
            url: url,
            method: options.method,
            body: options.body,
        });
    })
        .then(function (response) {
        try {
            return wrap_scryfall_response_1.default(response);
        }
        catch (err) {
            return Promise.reject(new scryfall_error_1.default({
                message: "Something went wrong when wrapping the response from Scryfall",
                thrownError: err,
            }));
        }
    })
        .catch(function (err) {
        if (!(err instanceof scryfall_error_1.default)) {
            err = new scryfall_error_1.default({
                message: "An unexpected error occurred when requesting resources from Scryfall.",
                status: err.status,
                originalError: err,
            });
        }
        return Promise.reject(err);
    });
}
exports.default = sendRequestToApi;
